presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -14,13 +14,15 @@
 package com.facebook.presto.sql.planner.iterative;
 
 import com.facebook.presto.Session;
+import com.facebook.presto.matching.Matchable;
+import com.facebook.presto.matching.Pattern;
 import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
 import com.facebook.presto.sql.planner.SymbolAllocator;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 
 import java.util.Optional;
 
-public interface Rule
+public interface Rule extends Matchable
```

Comment:
Please remove the `Notice that rule may be still invoked for plan nodes which given pattern does not apply, then rule should return Optional.empty() in such case` part of comment on `getPattern` - it's outdated. All the rules (that I've seen) declaring a pattern do the hard assumption of `ConcreteNode cNode = (ConcreteNode) node` in the very first line of `apply`. See `EvaluateZeroLimit` for an example. Also, I don't see a reason for the rules to be called with sth different than their pattern says - why would we need that?